packages/material-ui/src/ListItem/ListItem.js

```diff
@@ -104,72 +102,75 @@ const ListItem = React.forwardRef(function ListItem(props, ref) {
     ...other
   } = props;
 
-  return (
-    <MergeListContext dense={denseProp} alignItems={alignItems}>
-      {({ dense }) => {
-        const children = React.Children.toArray(childrenProp);
-        const hasAvatar = children.some(value => isMuiElement(value, ['ListItemAvatar']));
-        const hasSecondaryAction =
-          children.length &&
-          isMuiElement(children[children.length - 1], ['ListItemSecondaryAction']);
+  const context = React.useContext(ListContext);
+  const childContext = {
```

Comment:
Should probably memoize context values. Lists are usually the first candidates for perf issues.

Reply:
We have recently added memoization to the top-level List component. Does it worth the extra complexity for a leaf item? The potential harm is way smaller.

Reply:
Yeah let's just leave it as-is for now and wait for perf feedback.